### PR TITLE
Use cache for parsing files faster the second time

### LINKED ISSUE #876: Ability to cache parsed ASTs between semgrep-core runs
**Is your feature request related to a problem? Please describe.**
It would save unnecessary work in times when semgrep-core is called multiple times on the same file by semgrep to cache the AST representation somewhere instead of reparsing the file every time.

**Describe the solution you'd like**
Either a flag to pass to semgrep-core to use and clear a cache file or using a cache and semgrep clearing that cache after it is finished

### PR DESCRIPTION
**Pull request by aryx**
This should fix #876 (at least the ocaml part).
This saves parsed (and Naming_AST.resolved) generic AST
in /tmp/semgrep_core_cache<uid/ if the -use_parsing_cache command-line
argument was passed to semgrep-core.

To clean the cache, simply do
rm -f /tmp/semgrep_core_cache<uid/ from the Python side if you need to.

Test plan:
takes 2.1s the first time and 0.7sec the second time when testing
on tests/PERF/ js files.

/home/pad/github/semgrep/semgrep-core/_build/default/bin/Main.exe -equivalences /home/pad/github/semgrep/semgrep-core/data/basic_equivalences.yml -verbose -profile -e 1==2 -lang js tests/PERF/
---------------------
profiling result
---------------------
Main total                               :      1.996 sec          1 count
Parse_js.parse                           :      1.043 sec          4 count
Semgrep.check                            :      0.537 sec          4 count
Parse_js.tokens                          :      0.401 sec         13 count
Semgrep.match_e_e                        :      0.307 sec     132621 count
...

This time with -use_parsing_cache:
/home/pad/github/semgrep/semgrep-core/_build/default/bin/Main.exe -equivalences /home/pad/github/semgrep/semgrep-core/data/basic_equivalences.yml -use_parsing_cache -verbose -profile -e 1==2 -lang js tests/PERF/

---------------------
profiling result
---------------------
Main total                               :      0.546 sec          1 count
Semgrep.check                            :      0.486 sec          4 count
Semgrep.match_e_e                        :      0.285 sec     132621 count
rule:-e/-f                               :      0.105 sec     132621 count
Apply_equivalences.apply                 :      0.042 sec          4 count
Main.cache_computation                   :      0.040 sec          4 count
...

ls -al /tmp/semgrep_core_cache_1000/
total 9984
drwx------  2 pad  pad     4096 Jun  2 16:08 .
drwxrwxrwt 17 root root   49152 Jun  2 16:26 ..
-rw-r--r--  1 pad  pad  1019805 Jun  2 16:07 027e133a90b2c79d01c3224dfb0fed8f.ast_cache
-rw-r--r--  1 pad  pad  3953523 Jun  2 16:05 20d25efa622e4bfbc4015d03ba0d1d68.ast_cache
-rw-r--r--  1 pad  pad  3953523 Jun  2 16:08 6c05be109ca532cf545c082bda4a07fd.ast_cache
-rw-r--r--  1 pad  pad     7590 Jun  2 16:05 8cae6609c9048056880aa68cc383912c.ast_cache
-rw-r--r--  1 pad  pad     7590 Jun  2 16:07 92a4a1e679734b50b6dbc1cbfae37e86.ast_cache
-rw-r--r--  1 pad  pad    98103 Jun  2 16:07 a51fb3d6d3645f7004df0cee7b18e176.ast_cache
-rw-r--r--  1 pad  pad    98103 Jun  2 16:05 bd7d534ea53a44d782902098c97fedb9.ast_cache
-rw-r--r--  1 pad  pad  1019805 Jun  2 16:05 e763c7f5ef9424a2a7e295dd564cfabb.ast_cache

There are each time 2 files with the same size because I bumped ast_version
and so it regenerated new cache entries.

### CODE REVIEW

File: semgrep-core/bin/Main.ml [hunk 1]
```diff
@@ -264,6 +275,55 @@ let gen_layer ~root ~query file =
...
+ * This function is slightly more flexible because we can put the cache file
+ * anywhere thanks to the argument 'cache_file_of_file'.
+ * We also try to be a bit more type-safe by using the version tag above.
+ * TODO: merge in pfff/commons/Common.ml at some point
+ *)
+let cache_computation file cache_file_of_file f =
+  if not !use_parsing_cache
+  then f ()
+  else begin
+    if not (Sys.file_exists file)
```
Comment (nbrahms) [inline semgrep-core/bin/Main.ml:283]: What are your thoughts re. adding the file md5 checksum here? This would enable the Python wrapper to skip calculating this value itself.

File: semgrep-core/bin/Main.ml [hunk 2]
```diff
@@ -264,6 +275,55 @@ let gen_layer ~root ~query file =
...
+    end
+  end
+
+
+let cache_file_of_file filename =
+  let dir = spf "/tmp/semgrep_core_cache_%d" (Unix.getuid()) in
+  if not (Sys.file_exists dir)
+  then Unix.mkdir dir 0o700;
+  (* hopefully there will be no collision *)
+  let md5 = Digest.string filename in
```
Comment (nbrahms) [inline semgrep-core/bin/Main.ml:323]: What's the purpose of hashing the filename here?

File: semgrep-core/bin/Main.ml [hunk 3]
```diff
@@ -264,6 +275,55 @@ let gen_layer ~root ~query file =
...
+    if not (Sys.file_exists file)
+    then begin
+      pr2 ("WARNING: cache_computation: can't find file "  ^ file);
+      pr2 ("defaulting to calling the function");
+      f ()
+    end else begin
+    profile_code "Main.cache_computation" (fun () ->
+
+      let file_cache = cache_file_of_file file in
+      if Sys.file_exists file_cache && filemtime file_cache >= filemtime file
```
Comment (nbrahms) [inline semgrep-core/bin/Main.ml:292]: Not sure how Docker cp works, but just flagging that this computation might always return false for the Docker container used with Docker CP.

File: semgrep-core/bin/Main.ml (hunk 2 above)
Comment (aryx) [reply semgrep-core/bin/Main.ml:323]: Just to create easily a filename in /tmp/semgrep_core_cache<uid>/ that correspond to a complex path. semgrep will pass me  /path/to/project/nested/dir/foo.c  and I will hash that.
I could also replace all / with __, but I might get escaping issues, so just using the md5sum of the path seemed simpler. Also it prevents people to play with the cache directory. Those filename are a bit more opaque than a /tmp/semgrep_core/path__to__project__nested__dir__foo_c

File: semgrep-core/bin/Main.ml (hunk 1 above)
Comment (aryx) [reply semgrep-core/bin/Main.ml:283]: we don't compute the md5sum of the file content, just of the filename

Comment (nbrahms) [reply semgrep-core/bin/Main.ml:283]: The suggestion here is that we do calculate the md5 sum of the file content. Given that we're already lexing the entire file, this seems relatively cheap for us here, vs. done in the Python wrapper.

File: semgrep-core/bin/Main.ml (hunk 2 above)
Comment (nbrahms) [reply semgrep-core/bin/Main.ml:323]: Gotcha. What does happen if there's a collision? Can we at least store the full file path in the cache data, along with the `NamedAST`? That way we can tell when we've hit a collision.

File: semgrep-core/bin/Main.ml [hunk 4]
```diff
@@ -136,6 +136,17 @@ let supported_langs: string = String.concat ", " keys
...
 (*e: constant [[Main_semgrep_core.ncores]] *)
 
+(* TODO: we may need to put that in AST_generic.ml at some point
+ * and people should bump this number each time they make a modification
+ * to the generic AST (or to some analysis such as Naming_AST.ml which
+ * will have an impact on the marshalled AST). Otherwise, we may
+ * get some segmentation fault as OCaml marshalling is not entirely
+ * type-safe!!
+ *)
+let ast_version = 2
```
Comment (nbrahms) [inline semgrep-core/bin/Main.ml]: I suggest that this be the full version (including git commit) of `pfff`, so that any rev to `pfff` invalidates the cache.

Comment (nbrahms) [reply semgrep-core/bin/Main.ml]: For development, forcing invalidation whenever `pfff` is dirty would also prevent developer pain.

**Comment by mschwager** [comment]
Has there been any higher level discussion on the trade-offs between complexity and speed here? Caching can be simple at first glance, but difficult to get _completely_ correct, and can make debugging much more difficult. I.e. "why are we seeing these results? oh, there's some complex interaction with caching going on." <- this can be a pain getting right on the multitude of systems and environments we expect `semgrep` to run on.

For example, it seems there's already a lot of discussion around what the caching key should be. It looks like we currently think it should include:

* File path - can there be collisions here?
* Language, e.g. `--lang`
* `pfff` version
* A manually set AST version that we're asking developers to remember to bump at certain times

Is there anything we're missing? Are we painting ourselves in a corner right now with design decisions we may regret?

I'm also curious, what happens if `/tmp` doesn't exist? There are weird systems out there. This seems to depend on `Unix.getuid()` vs. something like `NamedTemporaryFile` in Python, which feels fragile. Are there collisions here we may not expect? How do we debug this functionality on an external user's system? Is it easier to implement this in Python vs. OCaml? E.g. hashing CLI args passed to `semgrep-core` + a hash of the file contents _should_ always produce the same results. Does this work with YAML URL functionality, or anything else that can change out from underneath us?

Basically, I'm trying to highlight the complexity of the solution, and implementing caching correctly in general. My vote would be to pass on this optimization until we see a greater need for it, or until we come up with a peer-reviewed design we believe to be rock-solid. We've already seen great speedups recently, and while 2.1s -> 0.7s is another great speedup, it doesn't warrant the additional complexity, IMO.

See also some discussion on similar functionality in `flake8`: https://gitlab.com/pycqa/flake8/-/issues/601

File: semgrep-core/bin/Main.ml (hunk 1 above)
Comment (rcoh) [reply semgrep-core/bin/Main.ml:283]: I think caching on md5 sum most reasonable approach. Caching on filename / time which has weird issues among other things inside K8S based docker containers isn't worthwhile. 

I don't think in real world analysis that parsing seems to be a big part of runtime anyway?

File: semgrep-core/bin/Main.ml (hunk 2 above)
Comment (mjambon) [reply semgrep-core/bin/Main.ml:324]: Separate conversation about `/tmp`: you can use `Filename.get_temp_dir_name ()` instead of assuming `/tmp` exists. (this point was raised by @mschwager )

from https://caml.inria.fr/pub/docs/manual-ocaml/libref/Filename.html :
```
val get_temp_dir_name : unit -> string
```
> The name of the temporary directory: Under Unix, the value of the TMPDIR environment variable, or "/tmp" if the variable is not set. Under Windows, the value of the TEMP environment variable, or "." if the variable is not set. The temporary directory can be changed with Filename.set_temp_dir_name.
>
> Since 4.00.0
```

**Comment by brendongo** [comment]
@mschwager good flag re: complexity vs performance here. Definitely share your sentiments on how hard it is to get caching right. We've definitely been bit by caching bugs in bento. I'll hold of on merging this until #863 is ready so we can see what the performance hit that gets and what the performance gain of caching AST gives us.

File: semgrep-core/bin/Main.ml [hunk 5]
```diff
@@ -264,6 +275,55 @@ let gen_layer ~root ~query file =
...
+        Common2.write_value (ast_version, res) file_cache;
+        res
+      end
+      )
+    end
+  end
+
+
+let cache_file_of_file filename =
+  let dir = spf "/tmp/semgrep_core_cache_%d" (Unix.getuid()) in
```
Comment (mjambon) [inline semgrep-core/bin/Main.ml]: About `/tmp`: you can use `Filename.get_temp_dir_name ()` instead of assuming `/tmp` exists. (this point was raised by @mschwager )

from https://caml.inria.fr/pub/docs/manual-ocaml/libref/Filename.html :
```
val get_temp_dir_name : unit -> string
```
> The name of the temporary directory: Under Unix, the value of the TMPDIR environment variable, or "/tmp" if the variable is not set. Under Windows, the value of the TEMP environment variable, or "." if the variable is not set. The temporary directory can be changed with Filename.set_temp_dir_name.
>
> Since 4.00.0
```

**Comment by mjambon** [comment]
I agree with the concerns about getting caching right in general, but I'm not in a position to say whether this should be done here. It's good to have a flag to enable/disable caching, and there should be tests to check they produce the same results.

**Comment by mjambon** [comment]
Things that may have an impact: is the cached AST sensitive to command-line options or config files? (how many, is it going to change?) or worse, environment variables? Depending on how messy that is, caching could be a breeze or a nightmare.

File: semgrep-core/bin/Main.ml (hunk 2 above)
Comment (aryx) [reply semgrep-core/bin/Main.ml:323]: Good idea!

File: semgrep-core/bin/Main.ml (hunk 4 above)
Comment (aryx) [reply semgrep-core/bin/Main.ml]: Sgtm. I think brendon generate a version.ml that we can use here

File: semgrep-core/bin/Main.ml (hunk 5 above)
Comment (aryx) [reply semgrep-core/bin/Main.ml]: Good idea!

**Comment by aryx** [comment]
Yes the solution here is small and easy to undo. I also would prefer no caching, but it can be useful to experiment and see how much it simplifies semgrep-python when you can call multiple times semgrepcore

**Comment by nbrahms** [comment]
As long as the caching isn't used by default, it makes sense to me to integrate it now so we can start playing with it.
